File: facecat_cs/grid/FCGridRow.cs
/*捂脸猫FaceCat框架 v1.0
 1.创始人-矿洞程序员-上海宁米科技创始人-脉脉KOL-陶德 (微信号:suade1984);
 2.联合创始人-上海宁米科技创始人-袁立涛(微信号:wx627378127);
 3.联合创始人-河北思尔企业管理咨询有限公司合伙人-肖添龙(微信号:xiaotianlong_luu);
 4.联合开发者-陈晓阳(微信号:chenxiaoyangzxy)，助理-朱炜(微信号:cnnic_zhu);
 5.该框架开源协议为BSD，欢迎对我们的创业活动进行各种支持，欢迎更多开发者加入。
 包含C/C++,Java,C#,iOS,MacOS,Linux六个版本的图形和通讯服务框架。
 */
using System;
using System.Collections.Generic;

namespace FaceCat {
    /// <summary>
    /// 表格行的样式
    /// </summary>
    public class FCGridRowStyle {
        protected long m_backColor = FCColor.Back;

        /// <summary>
        /// 获取
        /// </summary>
        public virtual long BackColor {
            get { return m_backColor; }
            set { m_backColor = value; }
        }

        protected FCFont m_font = new FCFont("Simsun", 14, false, false, false);

        /// <summary>
        /// 获取或设置字体
        /// </summary>
        public virtual FCFont Font {
            get { return m_font; }
            set { m_font = value; }
        }

        protected long m_hoveredBackColor = FCColor.argb(150, 150, 150);

        /// <summary>
        /// 获取或设置触摸悬停行的背景色
        /// </summary>
        public virtual long HoveredBackColor {
            get { return m_hoveredBackColor; }
            set { m_hoveredBackColor = value; }
        }

        protected long m_hoveredTextColor = FCColor.Text;

        /// <summary>
        /// 获取或设置触摸悬停行的前景色
        /// </summary>
        public virtual long HoveredTextColor {
            get { return m_hoveredTextColor; }
            set { m_hoveredTextColor = value; }
        }

        protected long m_selectedBackColor = FCColor.argb(100, 100, 100);

        /// <summary>
        /// 获取或设置选中行的背景色
        /// </summary>
        public virtual long SelectedBackColor {
            get { return m_selectedBackColor; }
            set { m_selectedBackColor = value; }
        }

        protected long m_selectedTextColor = FCColor.Text;

        /// <summary>
        /// 获取或设置选中行的前景色
        /// </summary>
        public virtual long SelectedTextColor {
            get { return m_selectedTextColor; }
            set { m_selectedTextColor = value; }
        }

        protected long m_textColor = FCColor.Text;

        /// <summary>
        /// 获取或设置前景色
        /// </summary>
        public virtual long TextColor {
            get { return m_textColor; }
            set { m_textColor = value; }
        }
    }

    /// <summary>
    /// 单元格
    /// </summary>
    public class FCGridRow : FCProperty {
        /// <summary>
        /// 创建行
        /// </summary>
        public FCGridRow() {
        }

        /// <summary>
        /// 析构函数
        /// </summary>
        ~FCGridRow() {
            delete();
        }

        /// <summary>
        /// 所有单元格
        /// </summary>
        public ArrayList<FCGridCell> m_cells = new ArrayList<FCGridCell>();

        /// <summary>
        /// 编辑状态
        /// </summary>
        public int m_editState;

        protected bool m_allowEdit;

        /// <summary>
        /// 获取或设置是否允许编辑
        /// </summary>
        public virtual bool AllowEdit {
            get { return m_allowEdit; }
            set { m_allowEdit = value; }
        }

        protected FCRect m_bounds;

        /// <summary>
        /// 获取或设置显示区域
        /// </summary>
        public virtual FCRect Bounds {
            get { return m_bounds; }
            set { m_bounds = value; }
        }

        protected FCView m_editButton;

        /// <summary>
        /// 获取或设置编辑按钮
        /// </summary>
        public virtual FCView EditButton {
            get { return m_editButton; }
            set { m_editButton = value; }
        }

        protected FCGrid m_grid;

        /// <summary>
        /// 获取或设置所在表格
        /// </summary>
        public virtual FCGrid Grid {
            get { return m_grid; }
            set { m_grid = value; }
        }

        protected int height = 20;

        /// <summary>
        /// 获取或设置行高
        /// </summary>
        public virtual int Height {
            get { return height; }
            set { height = value; }
        }

        protected int m_horizontalOffset = 0;

        /// <summary>
        /// 获取或设置横向滚动距离
        /// </summary>
        public virtual int HorizontalOffset {
            get { return m_horizontalOffset; }
            set { m_horizontalOffset = value; }
        }

        protected int m_index = -1;

        /// <summary>
        /// 获取或设置索引
        /// </summary>
        public virtual int Index {
            get { return m_index; }
            set { m_index = value; }
        }

        protected bool m_isDeleted;

        /// <summary>
        /// 获取或设置是否已被销毁
        /// </summary>
        public bool IsDeleted {
            get { return m_isDeleted; }
        }

        protected object m_tag = null;

        /// <summary>
        /// 获取或设置TAG值
        /// </summary>
        public virtual object Tag {
            get { return m_tag; }
            set { m_tag = value; }
        }

        protected bool m_visible = true;

        /// <summary>
        /// 获取或设置是否可见
        /// </summary>
        public virtual bool Visible {
            get { return m_visible; }
            set { m_visible = value; }
        }

        protected int m_visibleIndex = -1;

        /// <summary>
        /// 获取或设置可见索引
        /// </summary>
        public virtual int VisibleIndex {
            get { return m_visibleIndex; }
            set { m_visibleIndex = value; }
        }

        /// <summary>
        /// 添加单元格
        /// </summary>
        /// <param name="column">列索引</param>
        /// <param name="cell">单元格</param>
        public void addCell(FCGridColumn column, FCGridCell cell) {
            cell.Grid = m_grid;
            cell.Column = column;
            cell.Row = this;
            m_cells.add(cell);
            cell.onAdd();
        }

        /// <summary>
        /// 添加单元格
        /// </summary>
        /// <param name="columnIndex">列索引</param>
        /// <param name="cell">单元格</param>
        public void addCell(int columnIndex, FCGridCell cell) {
            cell.Grid = m_grid;
            cell.Column = m_grid.getColumn(columnIndex);
            cell.Row = this;
            m_cells.add(cell);
            cell.onAdd();
        }

        /// <summary>
        /// 添加单元格
        /// </summary>
        /// <param name="columnName">列名</param>
        /// <param name="cell">单元格</param>
        public void addCell(String columnName, FCGridCell cell) {
            cell.Grid = m_grid;
            cell.Column = m_grid.getColumn(columnName);
            cell.Row = this;
            m_cells.add(cell);
            cell.onAdd();
        }

        /// <summary>
        /// 清除单元格
        /// </summary>
        public void clearCells() {
            int cellSize = m_cells.size();
            for (int i = 0; i < cellSize; i++) {
                m_cells.get(i).onRemove();
                m_cells.get(i).delete();
            }
            m_cells.clear();
        }

        /// <summary>
        /// 销毁对象
        /// </summary>
        public void delete() {
            if (!m_isDeleted) {
                m_editButton = null;
                m_isDeleted = !m_isDeleted;
                int cellSize = m_cells.size();
                for (int i = 0; i < cellSize; i++) {
                    m_cells.get(i).delete();
                }
                m_cells.clear();
            }
        }

        /// <summary>
        /// 根据列获取单元格
        /// </summary>
        /// <param name="column">列</param>
        /// <returns>单元格</returns>
        public FCGridCell getCell(FCGridColumn column) {
            int cellSize = m_cells.size();
            for (int i = 0; i < cellSize; i++) {
                if (m_cells.get(i).Column == column) {
                    return m_cells.get(i);
                }
            }
            return null;
        }

        /// <summary>
        /// 根据列的索引获取单元格
        /// </summary>
        /// <returns>单元格</returns>
        public FCGridCell getCell(int columnIndex) {
            int cellsSize = m_cells.size();
            if (cellsSize > 0) {
                if (columnIndex >= 0 && columnIndex < cellsSize) {
                    if (m_cells.get(columnIndex).Column.Index == columnIndex) {
                        return m_cells.get(columnIndex);
                    }
                }
                for (int i = 0; i < cellsSize; i++) {
                    if (m_cells.get(i).Column.Index == columnIndex) {
                        return m_cells.get(i);
                    }
                }
            }
            return null;
        }

        /// <summary>
        /// 根据列的名称获取单元格
        /// </summary>
        /// <param name="columnName">列名</param>
        /// <returns>单元格</returns>
        public FCGridCell getCell(String columnName) {
            int cellSize = m_cells.size();
            for (int i = 0; i < cellSize; i++) {
                if (m_cells.get(i).Column.Name == columnName) {
                    return m_cells.get(i);
                }
            }
            return null;
        }

        /// <summary>
        /// 获取所有单元格
        /// </summary>
        /// <returns>所有单元格</returns>
        public ArrayList<FCGridCell> getCells() {
            return m_cells;
        }

        /// <summary>
        /// 获取属性值
        /// </summary>
        /// <param name="name">属性名称</param>
        /// <param name="value">返回属性值</param>
        /// <param name="type">返回属性类型</param>
        public virtual void getProperty(String name, ref String value, ref String type) {
            if (name == "allowedit") {
                type = "bool";
                value = FCStr.convertBoolToStr(AllowEdit);
            }
            else if (name == "height") {
                type = "int";
                value = FCStr.convertIntToStr(Height);
            }
            else if (name == "visible") {
                type = "int";
                value = FCStr.convertBoolToStr(Visible);
            }
            else {
                type = "undefined";
                value = "";
            }
        }

        /// <summary>
        /// 获取属性名称列表
        /// </summary>
        /// <returns></returns>
        public virtual ArrayList<String> getPropertyNames() {
            ArrayList<String> propertyNames = new ArrayList<String>();
            propertyNames.AddRange(new String[] { "AllowEdit", "EditorWidth", "Height", "Visible" });
            return propertyNames;
        }

        /// <summary>
        /// 添加行方法
        /// </summary>
        public virtual void onAdd() {
            int cellSize = m_cells.size();
            for (int i = 0; i < cellSize; i++) {
                m_cells.get(i).onAdd();
            }
        }


        /// <summary>
        /// 重绘方法
        /// </summary>
        /// <param name="paint">绘图对象</param>
        /// <param name="clipRect">裁剪矩形</param>
        /// <param name="isAlternate">是否交替行</param>
        public virtual void onPaint(FCPaint paint, FCRect clipRect, bool isAlternate) {

        }

        /// <summary>
        /// 重绘边线方法
        /// </summary>
        /// <param name="paint">绘图对象</param>
        /// <param name="clipRect">裁剪矩形</param>
        /// <param name="isAlternate">是否交替行</param>
        public virtual void onPaintBorder(FCPaint paint, FCRect clipRect, bool isAlternate) {

        }

        /// <summary>
        /// 移除行方法
        /// </summary>
        public virtual void onRemove() {
            int cellSize = m_cells.size();
            for (int i = 0; i < cellSize; i++) {
                m_cells.get(i).onRemove();
            }
        }

        /// <summary>
        /// 移除单元格
        /// </summary>
        /// <param name="column">列</param>
        public void removeCell(FCGridColumn column) {
            int cellSize = m_cells.size();
            for (int i = 0; i < cellSize; i++) {
                FCGridCell cell = m_cells.get(i);
                if (cell.Column == column) {
                    m_cells.remove(cell);
                    cell.onRemove();
                    break;
                }
            }
        }

        /// <summary>
        /// 移除单元格
        /// </summary>
        /// <param name="columnIndex">列索引</param>
        public void removeCell(int columnIndex) {
            int cellSize = m_cells.size();
            if (columnIndex >= 0 && columnIndex < cellSize) {
                FCGridCell cell = m_cells.get(columnIndex);
                if (cell.Column.Index == columnIndex) {
                    m_cells.remove(cell);
                    cell.onRemove();
                    return;
                }
                for (int i = 0; i < cellSize; i++) {
                    cell = m_cells.get(i);
                    if (cell.Column.Index == columnIndex) {
                        m_cells.remove(cell);
                        cell.onRemove();
                        break;
                    }
                }
            }
        }

        /// <summary>
        /// 移除单元格
        /// </summary>
        /// <param name="columnName">列名</param>
        public void removeCell(String columnName) {
            int cellSize = m_cells.size();
            for (int i = 0; i < cellSize; i++) {
                FCGridCell cell = m_cells.get(i);
                if (cell.Column.Name == columnName) {
                    m_cells.remove(cell);
                    cell.onRemove();
                    break;
                }
            }
        }

        /// <summary>
        /// 设置属性
        /// </summary>
        /// <param name="name">属性名称</param>
        /// <param name="value">属性值</param>
        public virtual void setProperty(String name, String value) {
            if (name == "allowedit") {
                AllowEdit = FCStr.convertStrToBool(value);
            }
            else if (name == "height") {
                Height = FCStr.convertStrToInt(value);
            }
            else if (name == "visible") {
                Visible = FCStr.convertStrToBool(value);
            }
        }
    }
}

File: facecat_cpp/source/grid/FCGridRow.cpp
#include "..\\..\\stdafx.h"
#include "..\\..\\include\\grid\\FCGridRow.h"

namespace FaceCat{
	FCGridRowStyle::FCGridRowStyle(){
		m_backColor = FCColor_Back;
		m_font = new FCFont(L"Simsun", 14, false, false, false);;
		m_textColor = FCColor_Text;
		m_hoveredBackColor = FCColor::argb(150, 150, 150);
		m_hoveredTextColor = FCColor_Text;
		m_selectedBackColor = FCColor::argb(100, 100, 100);
		m_selectedTextColor = FCColor_Text;
	}

	FCGridRowStyle::~FCGridRowStyle(){
		if(m_font){
			delete m_font;
			m_font = 0;
		}
	}

	Long FCGridRowStyle::getBackColor(){
		return m_backColor;
	}

	void FCGridRowStyle::setBackColor(Long backColor){
		m_backColor = backColor;
	}

	FCFont* FCGridRowStyle::getFont(){
		return m_font;
	}

	void FCGridRowStyle::setFont(FCFont *font){
		if(font){
			if(!m_font){
				m_font = new FCFont();
			}
			m_font->copy(font);
		}
		else{
			if(m_font){
				delete m_font;
				m_font = 0;
			}
		}
	}

	Long FCGridRowStyle::getHoveredBackColor(){
		return m_hoveredBackColor;
	}

	void FCGridRowStyle::setHoveredBackColor(Long hoveredBackColor){
		m_hoveredBackColor = hoveredBackColor;
	}

	Long FCGridRowStyle::getHoveredTextColor(){
		return m_hoveredTextColor;
	}

	void FCGridRowStyle::setHoveredTextColor(Long hoveredTextColor){
		m_hoveredTextColor = hoveredTextColor;
	}

	Long FCGridRowStyle::getSelectedBackColor(){
		return m_selectedBackColor;
	}

	void FCGridRowStyle::setSelectedBackColor(Long selectedBackColor){
		m_selectedBackColor = selectedBackColor;
	}

	Long FCGridRowStyle::getSelectedTextColor(){
		return m_selectedTextColor;
	}

	void FCGridRowStyle::setSelectedTextColor(Long selectedTextColor){
		m_selectedTextColor = selectedTextColor;
	}

	Long FCGridRowStyle::getTextColor(){
		return m_textColor;
	}

	void FCGridRowStyle::setTextColor(Long textColor){
		m_textColor = textColor;
	}

	void FCGridRowStyle::copy(FCGridRowStyle *style){
		setBackColor(style->getBackColor());
		setFont(style->getFont());
		setTextColor(style->getTextColor());
		setHoveredBackColor(style->getHoveredBackColor());
		setHoveredTextColor(style->getHoveredTextColor());
		setSelectedBackColor(style->getSelectedBackColor());
		setSelectedTextColor(style->getSelectedTextColor());
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	FCGridRow::FCGridRow(){
		m_allowEdit = false;
		m_bounds.left = 0;
		m_bounds.top = 0;
		m_bounds.right = 0;
		m_bounds.bottom = 0;
		m_editButton = 0;
		m_editState = 0;
		m_grid = 0;
		m_height = 20;
		m_horizontalOffset = 0;
		m_index = -1;
		m_tag = 0;
		m_visible = true;
		m_visibleIndex = -1;
	}

	FCGridRow::~FCGridRow(){
		for(int i = 0; i < m_cells.size(); i++){
			delete m_cells.get(i);
		}
		m_cells.clear();
		m_editButton = 0;
		m_grid = 0;
		m_tag = 0;
	}

	bool FCGridRow::allowEdit(){
		return m_allowEdit;
	}

	void FCGridRow::setAllowEdit(bool allowEdit){
		m_allowEdit = allowEdit;
	}

	FCRect FCGridRow::getBounds(){
		return m_bounds;
	}

	void FCGridRow::setBounds(FCRect bounds){
		m_bounds = bounds;
	}

	FCView* FCGridRow::getEditButton(){
		return m_editButton;
	}

	void FCGridRow::setEditButton(FCView *editButton){
		m_editButton = editButton;
	}

	FCGrid* FCGridRow::getGrid(){
		return m_grid;
	}

	void FCGridRow::setGrid(FCGrid *grid){
		m_grid = grid;
	}

	int FCGridRow::getHeight(){
		return m_height;
	}

	void FCGridRow::setHeight(int height){
		m_height = height;
	}

	int FCGridRow::getHorizontalOffset(){
		return m_horizontalOffset;
	}

	void FCGridRow::setHorizontalOffset(int horizontalOffset){
		m_horizontalOffset = horizontalOffset;
	}

	int FCGridRow::getIndex(){
		return m_index;
	}

	void FCGridRow::setIndex(int index){
		m_index = index;
	}

	Object FCGridRow::getTag(){
		return m_tag;
	}

	void FCGridRow::setTag(Object tag){
		m_tag = tag;
	}

	bool FCGridRow::isVisible(){
		return m_visible;
	}

	void FCGridRow::setVisible(bool visible){
		m_visible = visible;
	}

	int FCGridRow::getVisibleIndex(){
		return m_visibleIndex;
	}

	void FCGridRow::setVisibleIndex(int visibleIndex){
		m_visibleIndex = visibleIndex;
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	void FCGridRow::addCell(FCGridColumn *column, FCGridCell *cell){
		cell->setGrid(m_grid);
		cell->setColumn(column);
		cell->setRow(this);
		m_cells.add(cell);
		cell->onAdd();
	}

	void FCGridRow::addCell(const String& columnName, FCGridCell *cell){
		cell->setGrid(m_grid);
		cell->setColumn(m_grid->getColumn(columnName));
		cell->setRow(this);
		m_cells.add(cell);
		cell->onAdd();
	}


	void FCGridRow::addCell(int columnIndex, FCGridCell *cell){
		cell->setGrid(m_grid);
		cell->setColumn(m_grid->getColumn(columnIndex));
		cell->setRow(this);
		m_cells.add(cell);
		cell->onAdd();
	}

	void FCGridRow::clearCells(){
		for(int i = 0; i < m_cells.size(); i++){
			m_cells.get(i)->onRemove();
			delete m_cells.get(i);
		}
		m_cells.clear();
	}

	FCGridCell* FCGridRow::getCell(FCGridColumn *column){
		for(int i = 0; i < m_cells.size(); i++){
			FCGridCell *cell = m_cells.get(i);
			if(cell->getColumn() == column){
				return cell;
			}
		}
		return 0;
	}

	FCGridCell* FCGridRow::getCell(int columnIndex){
		int cellsSize = (int)m_cells.size();
		if(cellsSize > 0){
			if(columnIndex >= 0 && columnIndex < cellsSize){
				if(m_cells.get(columnIndex)->getColumn()->getIndex() == columnIndex){
					return m_cells.get(columnIndex);
				}
			}
			for(int i = 0; i < m_cells.size(); i++){
				FCGridCell *cell = m_cells.get(i);
				if(cell->getColumn()->getIndex() == columnIndex){
					return cell;
				}
			}
		}
		return 0;
	}

	FCGridCell* FCGridRow::getCell(const String& columnName){
		for(int i = 0; i < m_cells.size(); i++){
			FCGridCell *cell = m_cells.get(i);
			if(cell->getColumn()->getName() == columnName){
				return cell;
			}
		}
		return 0;
	}

	ArrayList<FCGridCell*> FCGridRow::getCells(){
		return m_cells;
	}

	void FCGridRow::getProperty(const String& name, String *value, String *type){
		if (name == L"allowedit"){
            *type = L"bool";
			*value = FCStr::convertBoolToStr(allowEdit());
        }
		else if (name == L"height"){
			*type = L"int";
            *value = FCStr::convertIntToStr(getHeight());
        }
        else if (name == L"visible"){
			*type = L"int";
            *value = FCStr::convertBoolToStr(isVisible());

        }
        else{
			*type = L"undefined";
            *value = L"";
        }
	}

	ArrayList<String> FCGridRow::getPropertyNames(){
		ArrayList<String> propertyNames;
		propertyNames.add(L"AllowEdit");
		propertyNames.add(L"Height");
		propertyNames.add(L"Visible");
		return propertyNames;
	}

	void FCGridRow::onAdd(){
		for(int i = 0; i < m_cells.size(); i++){
			m_cells.get(i)->onAdd();
		}
	}

	void FCGridRow::onPaint(FCPaint *paint, const FCRect& clipRect, bool isAlternate){
	}

	void FCGridRow::onPaintBorder(FCPaint *paint, const FCRect& clipRect, bool isAlternate){
	}

	void FCGridRow::onRemove(){
		for(int i = 0; i < m_cells.size(); i++){
			m_cells.get(i)->onRemove();
		}
	}

	void FCGridRow::removeCell(FCGridColumn *column){
		for(int i = 0; i < m_cells.size(); i++){
			FCGridCell *cell = m_cells.get(i);
			if(cell->getColumn() == column){
				m_cells.removeAt(i);
				cell->onRemove();
				break;
			}
		}
	}

	void FCGridRow::removeCell(int columnIndex){
		int cellSize = (int)m_cells.size();
		if(columnIndex >= 0 && columnIndex < cellSize){
			FCGridCell *cell = m_cells.get(columnIndex);
			if(cell->getColumn()->getIndex() == columnIndex){
				m_cells.removeAt(columnIndex);
				return;
			}
			for(int i = 0; i < m_cells.size(); i++){
				cell = m_cells.get(i);
				if(cell->getColumn()->getIndex() == columnIndex){
					m_cells.removeAt(i);
					cell->onRemove();
					break;
				}
			}
		}
	}

	void FCGridRow::removeCell(const String& columnName){
		for(int i = 0; i < m_cells.size(); i++){
			FCGridCell *cell = m_cells.get(i);
			if(cell->getColumn()->getName() == columnName){
				m_cells.removeAt(i);
				cell->onRemove();
				break;
			}
		}
	}

	void FCGridRow::setProperty(const String& name, const String& value){
		if (name == L"allowedit"){
			setAllowEdit(FCStr::convertStrToBool(value));
        }
		else if (name == L"height"){
            setHeight(FCStr::convertStrToInt(value));
        }
        else if (name == L"visible"){
            setVisible(FCStr::convertStrToBool(value));
        }
	}
}